src/Workspaces/CSharp/Portable/Workspace/LanguageServices/CSharpSyntaxTreeFactoryService.RecoverableSyntaxTree.cs

```diff
@@ -24,29 +24,33 @@ private partial class CSharpSyntaxTreeFactoryService
             /// the underlying full tree to stay alive.  Think of it more as a 
             /// key that can be used to identify a tree rather than the tree itself.
             /// </summary>
-            internal sealed class RecoverableSyntaxTree : CSharpSyntaxTree, IRecoverableSyntaxTree<CompilationUnitSyntax>, ICachedObjectOwner
+            internal sealed class RecoverableSyntaxTree : CSharpSyntaxTree, IRecoverableSyntaxTree<CSharpSyntaxNode>, ICachedObjectOwner
             {
-                private readonly RecoverableSyntaxRoot<CompilationUnitSyntax> _recoverableRoot;
+                private readonly RecoverableSyntaxRoot<CSharpSyntaxNode> _recoverableRoot;
                 private readonly SyntaxTreeInfo _info;
-                private readonly IProjectCacheHostService _projectCacheService;
+                private readonly AbstractSyntaxTreeFactoryService _service;
                 private readonly ProjectId _cacheKey;
 
+                public override bool HasCompilationUnitRoot { get; }
```

Comment:
What would be the effect of this change (i.e. allowing non CompilationUnit as root here)?

Reply:
Not sure what you mean. `WithRootAndOptions` accepts any syntax node as the root, so if we change the root of a recoverable tree to something that's not CompilationUnit we get a recoverable tree with that root.

Reply:
Sorry my question was quite vague. I'm wondering in what scenario this is needed, given that it doesn't seem to relate to the bug fix. Also consider covering this in the test

Reply:
Some features do this as I discovered when tests were failing. So it is covered by tests already :)